src/query/catalog/src/plan/internal_column.rs

```diff
@@ -0,0 +1,141 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_expression::types::string::StringColumnBuilder;
+use common_expression::types::DataType;
+use common_expression::types::NumberDataType;
+use common_expression::types::UInt64Type;
+use common_expression::BlockEntry;
+use common_expression::ColumnId;
+use common_expression::FromData;
+use common_expression::Scalar;
+use common_expression::TableDataType;
+use common_expression::Value;
+
+pub const ROW_ID: &str = "_row_id";
+pub const BLOCK_NAME: &str = "_block_name";
+pub const SEGMENT_NAME: &str = "_segment_name";
+pub const SNAPSHOT_NAME: &str = "_snapshot_name";
+
+// meta data for generate internal columns
+#[derive(Debug)]
+pub struct InternalColumnMeta {
+    pub segment_id: usize,
+    pub block_id: usize,
+    pub block_location: String,
+    pub segment_location: String,
+    pub snapshot_location: String,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq)]
+pub enum InternalColumnType {
+    RowId,
+    BlockName,
+    SegmentName,
+    SnapshotName,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq)]
+pub struct InternalColumn {
+    pub column_name: String,
+    pub column_type: InternalColumnType,
+}
+
+impl InternalColumn {
+    pub fn new(name: &str, column_type: InternalColumnType) -> Self {
+        InternalColumn {
+            column_name: name.to_string(),
+            column_type,
+        }
+    }
+
+    pub fn column_type(&self) -> &InternalColumnType {
+        &self.column_type
+    }
+
+    pub fn table_data_type(&self) -> TableDataType {
+        match &self.column_type {
+            InternalColumnType::RowId => TableDataType::Number(NumberDataType::UInt64),
+            InternalColumnType::BlockName => TableDataType::String,
+            InternalColumnType::SegmentName => TableDataType::String,
+            InternalColumnType::SnapshotName => TableDataType::String,
+        }
+    }
+
+    pub fn data_type(&self) -> DataType {
+        let t = &self.table_data_type();
+        t.into()
+    }
+
+    pub fn column_name(&self) -> &String {
+        &self.column_name
+    }
+
+    pub fn column_id(&self) -> ColumnId {
+        match &self.column_type {
+            InternalColumnType::RowId => u32::MAX,
+            InternalColumnType::BlockName => u32::MAX - 1,
+            InternalColumnType::SegmentName => u32::MAX - 2,
+            InternalColumnType::SnapshotName => u32::MAX - 3,
+        }
+    }
+
+    pub fn generate_column_values(&self, meta: &InternalColumnMeta, num_rows: usize) -> BlockEntry {
+        match &self.column_type {
+            InternalColumnType::RowId => {
+                let block_id = meta.block_id as u64;
+                let seg_id = meta.segment_id as u64;
+                let high_32bit = (seg_id << 48) + (block_id << 32);
```

Comment:
Num of Segment may be larger than `1<<16` (65536).

How about making `row_id =  { block_id  << 32 }    +  { block_row_id}`